### PR TITLE
Add futures to DeclarativeContent

### PR DESCRIPTION
**Pull request by mdellweg**
https://pulp.plan.io/issues/4294
re #4294

### CODE REVIEW

**Comment by mdellweg** [comment]
Keep in mind that the use of this needs a solution to https://pulp.plan.io/issues/4296 .
One benefit could be, that we see a content unit is in need of non blocking handling whenever it has a future.

**Comment by bmbouter** [comment]
@mdellweg this looks really good. I agree we need to find a solution for 4296 after this.

**Comment by mdellweg** [comment]
@bmbouter updated

File: pulpcore/plugin/stages/content_unit_stages.py [hunk 1]
```diff
@@ -184,3 +184,33 @@ class ContentUnitSaver(Stage):
 
         """
         pass
+
+
+class ResolveContentFutures(Stage):
+    """
+    This stage resolves the futures in `~pulpcore.plugin.stages.DeclarativeContent`.
```
Comment (bmbouter) [inline pulpcore/plugin/stages/content_unit_stages.py]: This isn't a merge blocker but if you put `:class:` in front of this sphinx makes it linkable.

File: pulpcore/plugin/stages/models.py [hunk 2]
```diff
@@ -61,16 +63,29 @@ class DeclarativeContent:
         d_artifacts (list): A list of zero or more
             :class:`~pulpcore.plugin.stages.DeclarativeArtifact` objects associated with `content`.
         extra_data (dict): A dictionary available for additional data to be stored in.
+        future (:class:`~asyncio.Future`): A future that gets resolved to the
+            `:class:~pulpcore.plugin.models.Content` in the
```
Comment (bmbouter) [inline pulpcore/plugin/stages/models.py]: Actually this won't work for Sphinx. the backtick needs to wrap the classpath and disinclude the `:class:` part.

**Comment by bmbouter** [comment]
@asmacdo what do you think about us merging this?

**Comment by asmacdo** [comment]
+1 to merging now. This should not affect any plugins until they are modified to use it.

**Pull request merged**